src/main/scala-2.11/markets/clearing/engines/CDAMatchingEngine.scala

```diff
@@ -15,24 +15,208 @@ limitations under the License.
+        if (matches.isEmpty) None else Some(matches)
+    }
+
+  }
+
+  @tailrec
+  private def findMatchingAskOrders(incoming: BidOrder,
+                                    matches: immutable.Queue[Match]): immutable.Queue[Match] = {
+    _askOrderBook.headOption match {
+      case Some(askOrder) if crosses(incoming, askOrder) =>
+
+        _askOrderBook -= askOrder  // SIDE EFFECT!
+        val residualQuantity = incoming.quantity - askOrder.quantity
+        val price = formPrice(incoming, askOrder)
+
+        if (residualQuantity < 0) {
+          val totalMatch = TotalMatch(askOrder, incoming, price)
+          // add residualOrder back into orderBook!
+          val residualOrder = askOrder.split(-residualQuantity)
+          _askOrderBook.add(residualOrder)  // SIDE EFFECT!
+          matches.enqueue(totalMatch)
+        } else if (residualQuantity == 0) {  // no rationing for incoming order!
+          val totalMatch = TotalMatch(askOrder, incoming, price)
+          matches.enqueue(totalMatch)
+        } else {  // incoming order is larger than existing order and will be rationed!
+          val partialMatch = PartialMatch(askOrder, incoming, price)
+          val residualOrder = incoming.split(residualQuantity)
+          findMatchingAskOrders(residualOrder, matches.enqueue(partialMatch))
+        }
+
+      case _ => // existingOrders is empty or incoming order does not cross best existing order.
+        _bidOrderBook.add(incoming)  // SIDE EFFECT!
+        matches
+    }
+  }
+
+  @tailrec
+  private def findMatchingBidOrders(incoming: AskOrder,
+                                    matches: immutable.Queue[Match]): immutable.Queue[Match] = {
+    _bidOrderBook.headOption match {
+      case Some(bidOrder) if crosses(incoming, bidOrder) =>
+
+        _bidOrderBook.remove(bidOrder)  // SIDE EFFECT!
+        val residualQuantity = incoming.quantity - bidOrder.quantity
+        val price = formPrice(incoming, bidOrder)
+
+        if (residualQuantity < 0) {
+          val totalMatch = TotalMatch(bidOrder, incoming, price)
+          // add residualOrder back into orderBook!
+          val residualOrder = bidOrder.split(-residualQuantity)
+          _bidOrderBook.add(residualOrder)  // SIDE EFFECT!
+          matches.enqueue(totalMatch)
+        } else if (residualQuantity == 0) {  // no rationing for incoming order!
+          val totalMatch = TotalMatch(bidOrder, incoming, price)
+          matches.enqueue(totalMatch)
+        } else {  // incoming order is larger than existing order and will be rationed!
+          val partialMatch = PartialMatch(bidOrder, incoming, price)
+          val residualOrder = incoming.split(residualQuantity)
+          findMatchingBidOrders(residualOrder, matches.enqueue(partialMatch))
+        }
+      case _ => // existingOrders is empty or incoming order does not cross best existing order.
+        _askOrderBook.add(incoming)  // SIDE EFFECT!
+        matches
+    }
+  }
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/davidrpugh/markets-sandbox/file/1519375750/issues/source?bid=2921361&fileBranchId=3008855#l135)